### DISCUSSION TITLE
Why is passStringToWasm0 implemented to reallocate twice when dealing with Unicode strings?

### DISCUSSION THREAD

**Discussion by clogic29** [wasm-bindgen/wasm-bindgen#4313] [Q&A]
Within the implementation of passStringToWasm0
```javascript
if (offset !== len) {
    if (offset !== 0) {
        arg = arg.slice(offset);
    }
    ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
    const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
    const ret = encodeString(arg, view);

    offset += ret.written;
    ptr = realloc(ptr, len, offset, 1) >>> 0;
}
```

In the implementation of passStringToWasm0, the ptr is reallocated twice:
Instead of reallocating arg.length * 3 initially, The length is determined after checking the written property of the view and then reallocated, And finally, the buf is appended to the ptr.

It seems possible to achieve this with only one reallocation. I'm curious about the considerations behind choosing an implementation with two reallocations.

**Comment by daxpedda** [comment]
I believe because we can't know beforehand how long its going to be after converting to UTF-8. So we allocate the maximum amount of memory its possibly going to take and then adjust it afterwards when we know how much it really took after the conversion.

If you have a suggestion how to do it only with one allocation that would be great!
